# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character_name：角色姓名。
- role：演职员类型，演员或导演。

## **1.讀取數據**

In [1]:
import pandas as pd

In [2]:
original_titles_data = pd.read_csv(r"C:\Users\User\Desktop\python\titles.csv")
original_titles_data.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [3]:
original_credits_data = pd.read_csv(r"C:\Users\User\Desktop\python\credits.csv")
original_credits_data.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

# 


## **2.評估數據**

### 2.1評估數據整齊度

#### 為了區分開經過清理的資料和原始的資料，我們創建新的變數`cleaned_titles_data`，讓它為`original_titles_data`複製出的副本，以及創建新的變數`cleaned_credits_data`，讓它為`original_credits_data`複製出的變數。我們之後的清理步驟都將被運用在`cleaned_titles_data`和`cleaned_credits_data`。

In [4]:
cleaned_titles_data = original_titles_data.copy()
cleaned_credits_data = original_credits_data.copy()

In [5]:
cleaned_titles_data.sample(10)

id                                     title   type  \
1434   ts53076                                  El Chema   SHOW   
2202   ts82631                              Monkey Twins   SHOW   
3932  tm474548                                  Good Sam  MOVIE   
1601  tm201855                               Katti Batti  MOVIE   
4364  ts225708                 Six Windows in the Desert   SHOW   
1532  tm311494                                      Slam  MOVIE   
1080  tm186408                         The Land of Cards  MOVIE   
726   ts191698                           Little Baby Bum   SHOW   
1056  tm152843                       Swearnet: The Movie  MOVIE   
5076  ts304053  Countdown: Inspiration4 Mission to Space   SHOW   

                                            description  release_year  \
1434  Inspired by true events, This is the story of ...          2016   
2202  Inspired by Khon dance drama and Thai martial ...          2018   
3932  A news reporter looks into who has been anonym...          2019   
1601  A man falls in love with a woman because of he...          2015   
4364  From a theatre play to the aftermath of a plan...          2020   
1532  Samuele is 16 and has a passion for skateboard...          2016   
1080  A royal prince arrives on an island of fascist...          2013   
726   Little Baby Bum is a fun and educational show ...          2011   
1056  Fed up with being censored in their post-Trail...          2014   
5076  From training to launch to landing, this all-a...          2021   

     age_certification  runtime                                      genres  \
1434             TV-MA       44                         ['crime', 'action']   
2202               NaN       49                         ['action', 'crime']   
3932                PG       90              ['romance', 'drama', 'comedy']   
1601               NaN      138              ['comedy', 'romance', 'drama']   
4364             TV-14       21     ['drama', 'scifi', 'comedy', 'history']   
1532               NaN      100  ['comedy', 'romance', 'drama', 'european']   
1080               NaN      110                                 ['fantasy']   
726               TV-Y        7            ['family', 'music', 'animation']   
1056                 R      112                                  ['comedy']   
5076             TV-14       50                           ['documentation']   

     production_countries  seasons     imdb_id  imdb_score  imdb_votes  \
1434               ['US']      1.0   tt5730624         6.9       308.0   
2202               ['TH']      1.0   tt8507498         7.9        95.0   
3932               ['US']      NaN   tt9032888         5.7      2951.0   
1601               ['IN']      NaN   tt4467262         4.6      3157.0   
4364               ['SA']      1.0  tt11874042         4.3      3104.0   
1532               ['IT']      NaN   tt4379536         6.1      1329.0   
1080               ['IN']      NaN   tt2261700         5.5       412.0   
726                ['GB']      4.0  tt10051854         6.4        94.0   
1056               ['CA']      NaN   tt2380564         6.0      2769.0   
5076               ['US']      1.0  tt15166116         7.9      1857.0   

      tmdb_popularity  tmdb_score  
1434          112.717         NaN  
2202            0.990         6.0  
3932            8.029         6.0  
1601            1.508         5.6  
4364            1.176        10.0  
1532           11.308         5.8  
1080            3.983         6.5  
726             0.971        10.0  
1056            4.778         5.4  
5076            6.054         6.7

In [6]:
cleaned_credits_data.sample(10)

person_id         id                name  \
59011    1831822   tm884616      Cristián Rojas   
53485    1002411   ts215718        Kim Dong-hee   
33658     154097   tm352774          Gay Talese   
58864     839182   tm372035    Clément Olivieri   
24860     286565   tm300713           Airi Kido   
69391    2042006  tm1028298  Slavik Pustovoytov   
28671      70729   tm288010       Peter Pedrero   
41858     597031   tm314272        Carol Portes   
41399       4893   tm324204         Katie Leigh   
44376       4108   tm244149       Steve Routman   

                                  character   role  
59011                      Security Guard 1  ACTOR  
53485                              Oh Jisoo  ACTOR  
33658                                  Self  ACTOR  
58864                                   NaN  ACTOR  
24860                              Ishikawa  ACTOR  
69391                       Jordan's Dancer  ACTOR  
28671                                 Anton  ACTOR  
41858                               Roberta  ACTOR  
41399      Carlotta / Patrick's Mom (voice)  ACTOR  
44376  Hoffa's Attorney - George Fitzgerald  ACTOR

#### 發現`cleaned_titles_data`的"genres"和"production_countries"的變數中包含多個值，應當進行拆分。

#### 而`cleaned_credits_data`符合評估數據整齊("每列一個變量，每行一個觀察值，每個單元格是一個值")。

# 


#### 首先我們先隨便查看`genres`一個變量進行觀察

In [7]:
cleaned_titles_data["genres"][1]

"['drama', 'crime']"

#### 結果顥示"['drama', 'crime']"，代表`genres`的變量不是字符串列表，而是字符串，無法直接用`value_counts`統計各個值出現的次數。 我們可以使用Python內置的`eval`函數，它可以把字串轉換成運算式，所以可以幫我們把表示清單的字串轉換成清單本身。

In [8]:
cleaned_titles_data["genres"] = cleaned_titles_data["genres"].apply(lambda x: eval(x))
cleaned_titles_data["genres"][1]

['drama', 'crime']

#### 轉換為列表後。就可以用DataFrame的`explode`的方法，把`genres`列表的列表拆分成單獨的行

In [9]:
cleaned_titles_data = cleaned_titles_data.explode("genres")
cleaned_titles_data.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

# 


#### 同樣的方法轉換`production_countries`

In [10]:
cleaned_titles_data["production_countries"][4]

4    ['GB', 'US']
4    ['GB', 'US']
Name: production_countries, dtype: object

#### `production_countries`的變量不是字符串列表，而是字符串，再次利用`eval`進行轉換列表類型

In [11]:
cleaned_titles_data["production_countries"] = cleaned_titles_data["production_countries"].apply(lambda x: eval(x))
cleaned_titles_data["production_countries"][1]

1    [US]
1    [US]
Name: production_countries, dtype: object

#### 用`explode`方法，把列表值拆分成單獨的行

In [12]:
cleaned_titles_data = cleaned_titles_data.explode("production_countries")
cleaned_titles_data.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   
2                 R      109       thriller                   US      NaN   
2                 R      109       european                   US      NaN   
3                PG       91        fantasy                   GB      NaN   
3                PG       91         action                   GB      NaN   
3                PG       91         comedy                   GB      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

# 


## **3.數據乾淨度**

### 3.1 先對`cleaned_titles_data`進行調整

In [13]:
cleaned_titles_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

#### 從輸出結果來看，`cleaned_titles_data`資料共有17818條觀察值，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`變數均存在缺失值，將在後續進行評估和清理。
#### 此外，`release_year`表示年份，資料類型不應為數字，應為日期，所以需要進行資料格式轉換。

# 



### 3.1.1把`release_year`的資料類型轉換為日期

In [14]:
cleaned_titles_data["release_year"] = pd.to_datetime(cleaned_titles_data["release_year"],format='%Y')
#cleaned_titles_data["release_year"] = cleaned_titles_data["release_year"].dt.date 
###意思：把 datetime 物件的「日期部分」抽出來（轉成 Python 的 datetime.date 物件）
###效果：時間 (00:00:00) 和微秒部分都會被去掉，只保留年月日。
cleaned_titles_data["release_year"]

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

# 


### 3.1.2處理缺失數據

#### 在`cleaned_titles_data`中，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`變數存在缺失值。

#### 由於影視作品的標題、描述、適齡認證、發行國家、電視劇季數、IMDB的ID、TMDB的流行度、TMDB的評分，並不影響我們挖掘各個流派中的高IMDB評分作品演員，所以可以保留`title`、`description`、`age_certification`、`production_countries`、`seasons`、`imdb_id`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`變數值存在空缺的觀察值。
#### 但`imdb_score`和`genres`，即IMDB評分和流派，和我們後續要做的分析息息相關。
#### 先提取出`imdb_score`缺失觀察值進行查看。


In [15]:
cleaned_titles_data[cleaned_titles_data["imdb_score"].isnull()]

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

#### `imdb_score`共有842行是空值，我們把空值刪除，並檢查空值的總數來檢驗是否空值都刪除。

In [16]:
cleaned_titles_data = cleaned_titles_data.dropna(subset=["imdb_score"])
cleaned_titles_data["imdb_score"].isnull().sum()                                          

np.int64(0)

#### 提取出`genres`缺失觀察值進行查看。

In [17]:
cleaned_titles_data[cleaned_titles_data["genres"].isnull()]

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

#### 我們把`genres`空值刪除，並檢查空值的總數來檢驗是否空值都刪除。

In [18]:
cleaned_titles_data = cleaned_titles_data.dropna(subset=["genres"])
cleaned_titles_data["genres"].isnull().sum()  

np.int64(0)

#

### 3.1.3處理重複數據

#### 根據資料變數的含義以及內容來看，`cleaned_titles_data`裡不應該存在每個變數值都相同的觀察值，因此查看是否存在重複值。

In [19]:
cleaned_titles_data.duplicated().sum()

np.int64(0)

#### 輸出結果是0,代表不存在重複值。

# 


### 3.1.4處理不一致數據

#### 針對`cleaned_titles_data`，不一致資料可能存在於`genres`和`production_countries`變數中，我們將查看是否存在多個不同值指代同一流派，以及多個不同值指代同一國家的情況。
#### 首先查看`genres`

In [20]:
cleaned_titles_data["genres"].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

#### 從上面看出，genres列裡並不存在不一致資料，各個值都在指代不同的流派。

In [21]:
#cleaned_titles_data = cleaned_titles_data.query('genres != ""')  [cleaned_titles_data → 你的 DataFrame.] [query("條件字串") → 用字串的形式寫篩選條件] [genres != ""' → 條件：genres 欄位的值 不等於空字串]
cleaned_titles_data.query('genres == ""')  ##輸出"genres"是空值的數據

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

#### 接著`production_countries`

In [22]:
cleaned_titles_data["production_countries"].value_counts()

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
CU       1
LK       1
GT       1
AF       1
FO       1
Name: count, Length: 108, dtype: int64

#### 由於`value_counts`執行結果中有太多值，`Pandas`只會預設顯示開頭和結尾的一些值。要完整展示結果，可以把`display.max_rows`設置為`None`，即取消展示行數上限。
#### 但因為我們只是在當前調用`value_counts`時才需要看完整結果，所以可以結合`option_context`，只更改臨時上限。


In [23]:
with pd.option_context('display.max_row', None):
    print(cleaned_titles_data['production_countries'].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
ES          637
KR          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
HK          102
NL          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
IS           28
AT           28
NZ           27
LU           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

#### 從以上輸出結果來看，出品國家都用兩位的國家代碼來表示，除了裡面存在一個的Lebanon值。
#### Lebanon的國家代碼是LB，出現了70次，說明此處資料不一致。LB和Lebanon都在表示同一國家，需要進行統一。
#### 把`cleaned_titles_data`裡，`production_countries`的"LB"和"Lebanon"統一為LB，並檢查替換後是否還存在"LB"


In [24]:
cleaned_titles_data["production_countries"] = cleaned_titles_data["production_countries"].replace({"Lebanon":"LB"})
with pd.option_context('display.max_row', None):
    print(cleaned_titles_data['production_countries'].value_counts())

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
ES     637
KR     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
SA      28
AT      28
IS      28
NZ      27
LU      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
MA      15
UY      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
VN       8
MT       8
SU       7
CD       7
TN       7
LT       7
IR       7
GH       6
SN       6
AL       6
KE       6
IQ       5
MU       5
CY       5
TZ       4
SY       4
MC       4
IO       4
KN       4
GR       4
BD       3
BS       3
DZ       3
GL       3
AO       3
CM       3


#### 另外，裡面還存在空字串表示的國家代碼，並非有效資料。但由於出品國家並非分析所需的關鍵資訊，所以可以保留出品國家為空的觀察值。

In [25]:
cleaned_titles_data.query('production_countries == ""') ##檢查是否有空值

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

#

### 3.1.5處理無效或錯誤數據

#### 可以通過DataFrame的describe方法，對數值統計資訊進行快速瞭解。

In [26]:
original_titles_data.describe()

release_year      runtime      seasons   imdb_score    imdb_votes  \
count   5850.000000  5850.000000  2106.000000  5368.000000  5.352000e+03   
mean    2016.417094    76.888889     2.162868     6.510861  2.343938e+04   
std        6.937726    39.002509     2.689041     1.163826  9.582047e+04   
min     1945.000000     0.000000     1.000000     1.500000  5.000000e+00   
25%     2016.000000    44.000000     1.000000     5.800000  5.167500e+02   
50%     2018.000000    83.000000     1.000000     6.600000  2.233500e+03   
75%     2020.000000   104.000000     2.000000     7.300000  9.494000e+03   
max     2022.000000   240.000000    42.000000     9.600000  2.294231e+06   

       tmdb_popularity   tmdb_score  
count      5759.000000  5539.000000  
mean         22.637925     6.829175  
std          81.680263     1.170391  
min           0.009442     0.500000  
25%           2.728500     6.100000  
50%           6.821000     6.900000  
75%          16.590000     7.537500  
max        2274.044000    10.000000

#### 從以上統計資訊來看，`original_titles_data`裡不存在脫離現實意義的數值。

#


### 3.2 對`cleaned_credits_data`進行調整

In [27]:
cleaned_credits_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


#### 從輸出結果來看，`cleaned_credits_data`資料共有77801條觀察值，`character`變數存在缺失值，將在後續進行評估和清理。
#### 此外，`person_id`表示演職員ID，資料類型不應為數位，應為字串，所以需要進行資料格式轉換。

#


### 3.2.1把`person_id`的資料類型轉換為字串

In [28]:
cleaned_credits_data["person_id"] = cleaned_credits_data["person_id"].astype("str")
cleaned_credits_data["person_id"]

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

#


### 3.2.2處理缺失數據

#### `character`變數存在缺失值，角色名並不影響我們挖掘各個流派中的高IMDB評分作品演員，並且此變數缺失也有可能因為演職員類別是導演，沒有對應角色，因此可以保留character變數值存在空缺的觀察值。

#


### 3.2.3處理重複數據

#### 查看cleaned_credits_data數據是否存在重複值

In [29]:
cleaned_credits_data.duplicated().sum()

np.int64(0)

#### 結果表示並沒有重複值

#


### 3.2.4處理不一致數據

#### 針對`cleaned_credits_data`，不一致資料可能存在於`role`中，我們將查看是否存在多個不同值指代同一演職員類型的情況。

In [30]:
cleaned_credits_data["role"].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

#### 從以上輸出結果來看，`role`只有兩種可能的值，`ACTOR`或`DIRECTOR`，不存在不一致資料。我們可以把這列的類型轉換為`Category`，好處是比字串類型更節約記憶體空間，也能表明說值的類型有限。

In [31]:
cleaned_credits_data["role"] = cleaned_credits_data["role"].astype("category")
cleaned_credits_data["role"]

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

#


### 3.2.5處理無效或錯誤數據

#### `original_credits`由於不包含表示數值含義的變數，因此無需用`describe`檢查。

#


## **4.整理數據**

In [32]:
cleaned_titles_data

id                               title   type  \
1       tm84618                         Taxi Driver  MOVIE   
1       tm84618                         Taxi Driver  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
...         ...                                 ...    ...   
5846   tm898842                         C/O Kaadhal  MOVIE   
5847  tm1059008                             Lokillo  MOVIE   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   

                                            description release_year  \
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
...                                                 ...          ...   
5846  A heart warming film that explores the concept...   2021-01-01   
5847  A controversial TV host and comedian who has b...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   

     age_certification  runtime     genres production_countries  seasons  \
1                    R      114      drama                   US      NaN   
1                    R      114      crime                   US      NaN   
2                    R      109      drama                   US      NaN   
2                    R      109     action                   US      NaN   
2                    R      109   thriller                   US      NaN   
...                ...      ...        ...                  ...      ...   
5846               NaN      134      drama                  NaN      NaN   
5847               NaN       90     comedy                   CO      NaN   
5849               NaN        7     family                  NaN      1.0   
5849               NaN        7  animation                  NaN      1.0   
5849               NaN        7     comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5846  tt11803618         7.7       348.0              NaN         NaN  
5847  tt14585902         3.8        68.0           26.005       6.300  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[16970 rows x 15 columns]

In [33]:
cleaned_credits_data

person_id         id                name                character  \
0          3748    tm84618      Robert De Niro            Travis Bickle   
1         14658    tm84618        Jodie Foster            Iris Steensma   
2          7064    tm84618       Albert Brooks                      Tom   
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4         48933    tm84618     Cybill Shepherd                    Betsy   
...         ...        ...                 ...                      ...   
77796    736339  tm1059008    Adelaida Buscato                María Paz   
77797    399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798    373198  tm1059008         Inés Prieto                    Fanny   
77799    378132  tm1059008        Isabel Gaona                   Cacica   
77800   1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

#### 此次數據分析主要目標是整理不同流派影視作品，演員出演作品的平均IMDEB評分，從而挖掘出各個流派中高評分作品演品。

#### 為了能同時獲得流派與演員數據，我們需要把`cleaned_titles_data`和`cleaned_credits_data`，通過`id`作為鍵進行連接，因為兩個數據表中`id`都是影視作品ID。

In [34]:
credits_with_titles = pd.merge(cleaned_titles_data, cleaned_credits_data, on="id", how="inner")
credits_with_titles.head()

id        title   type  \
0  tm84618  Taxi Driver  MOVIE   
1  tm84618  Taxi Driver  MOVIE   
2  tm84618  Taxi Driver  MOVIE   
3  tm84618  Taxi Driver  MOVIE   
4  tm84618  Taxi Driver  MOVIE   

                                         description release_year  \
0  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  drama                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  drama                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score person_id  \
0         8.2    808582.0           40.965       8.179      3748   
1         8.2    808582.0           40.965       8.179     14658   
2         8.2    808582.0           40.965       8.179      7064   
3         8.2    808582.0           40.965       8.179      3739   
4         8.2    808582.0           40.965       8.179     48933   

              name                character   role  
0   Robert De Niro            Travis Bickle  ACTOR  
1     Jodie Foster            Iris Steensma  ACTOR  
2    Albert Brooks                      Tom  ACTOR  
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  
4  Cybill Shepherd                    Betsy  ACTOR

#### 由於我們只對挖掘演員的參演作品口碑感興趣，導演不在我們的分析範圍內，因此根據`role`，篩選出類型為`ACTOR`的觀察值，供後續分析。

In [35]:
actor_with_titles = credits_with_titles.query('role == "ACTOR"')
actor_with_titles

id        title   type  \
0         tm84618  Taxi Driver  MOVIE   
1         tm84618  Taxi Driver  MOVIE   
2         tm84618  Taxi Driver  MOVIE   
3         tm84618  Taxi Driver  MOVIE   
4         tm84618  Taxi Driver  MOVIE   
...           ...          ...    ...   
276103  tm1059008      Lokillo  MOVIE   
276104  tm1059008      Lokillo  MOVIE   
276105  tm1059008      Lokillo  MOVIE   
276106  tm1059008      Lokillo  MOVIE   
276107  tm1059008      Lokillo  MOVIE   

                                              description release_year  \
0       A mentally unstable Vietnam War veteran works ...   1976-01-01   
1       A mentally unstable Vietnam War veteran works ...   1976-01-01   
2       A mentally unstable Vietnam War veteran works ...   1976-01-01   
3       A mentally unstable Vietnam War veteran works ...   1976-01-01   
4       A mentally unstable Vietnam War veteran works ...   1976-01-01   
...                                                   ...          ...   
276103  A controversial TV host and comedian who has b...   2021-01-01   
276104  A controversial TV host and comedian who has b...   2021-01-01   
276105  A controversial TV host and comedian who has b...   2021-01-01   
276106  A controversial TV host and comedian who has b...   2021-01-01   
276107  A controversial TV host and comedian who has b...   2021-01-01   

       age_certification  runtime  genres production_countries  seasons  \
0                      R      114   drama                   US      NaN   
1                      R      114   drama                   US      NaN   
2                      R      114   drama                   US      NaN   
3                      R      114   drama                   US      NaN   
4                      R      114   drama                   US      NaN   
...                  ...      ...     ...                  ...      ...   
276103               NaN       90  comedy                   CO      NaN   
276104               NaN       90  comedy                   CO      NaN   
276105               NaN       90  comedy                   CO      NaN   
276106               NaN       90  comedy                   CO      NaN   
276107               NaN       90  comedy                   CO      NaN   

           imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  \
0        tt0075314         8.2    808582.0           40.965       8.179   
1        tt0075314         8.2    808582.0           40.965       8.179   
2        tt0075314         8.2    808582.0           40.965       8.179   
3        tt0075314         8.2    808582.0           40.965       8.179   
4        tt0075314         8.2    808582.0           40.965       8.179   
...            ...         ...         ...              ...         ...   
276103  tt14585902         3.8        68.0           26.005       6.300   
276104  tt14585902         3.8        68.0           26.005       6.300   
276105  tt14585902         3.8        68.0           26.005       6.300   
276106  tt14585902         3.8        68.0           26.005       6.300   
276107  tt14585902         3.8        68.0           26.005       6.300   

       person_id                name                character   role  
0           3748      Robert De Niro            Travis Bickle  ACTOR  
1          14658        Jodie Foster            Iris Steensma  ACTOR  
2           7064       Albert Brooks                      Tom  ACTOR  
3           3739       Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  
4          48933     Cybill Shepherd                    Betsy  ACTOR  
...          ...                 ...                      ...    ...  
276103    368473        Aída Morales                  Maritza  ACTOR  
276104    736339    Adelaida Buscato                María Paz  ACTOR  
276105    399499  Luz Stella Luengas             Karen Bayona  ACTOR  
276106    373198         Inés Prieto                    Fanny  ACTOR  
276107    378132        Isabel Gaona              

#### 為了挖掘出各個流派中的高IMDB評分作品演員，我們需要先根據流派和演員進行分組。

#### 對演員進行分組的時候，選擇的是用`person_id`而不是`name`變量，原因是名字容易出現錯拼或者重名的情況，演職員ID會比演員姓名更加準確地反映是哪位演員。

In [36]:
groupby_genres_and_person_id = actor_with_titles.groupby(["genres","person_id"])

#### 分組後，我們只需要對`imdb_score`的值進行聚合計算，因此只提取`imdb_score`變量，然後調用`mean`，來計算各個流派影視作品中，每位演員參演作品的平均IMDB評分。

In [37]:
imdb_score_groupby_genres_and_person_id = groupby_genres_and_person_id["imdb_score"].mean()
imdb_score_groupby_genres_and_person_id

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

#### 我們可以調用`reset_index`，對層次化索引進行重置，得到更加規整的DataFrame。

In [38]:
imdb_score_groupby_genres_and_person_id_df = imdb_score_groupby_genres_and_person_id.reset_index()
imdb_score_groupby_genres_and_person_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

#### 把`name`放在`imdb_score_groupby_genres_and_person_id_df`

#### 為了得到演員ID所對應的演員名字，我們可以和`cleaned_credits`這個DataFrame進行連接。這個DataFrame還有其它列，我們只需要得到`person_id`和`name`的對應，所以可以先提取出那兩列，並把重復行刪除。

In [40]:
actor_id_with_names = cleaned_credits_data[['person_id', 'name']].drop_duplicates()
actor_id_with_names

person_id                name
0          3748      Robert De Niro
1         14658        Jodie Foster
2          7064       Albert Brooks
3          3739       Harvey Keitel
4         48933     Cybill Shepherd
...         ...                 ...
77796    736339    Adelaida Buscato
77797    399499  Luz Stella Luengas
77798    373198         Inés Prieto
77799    378132        Isabel Gaona
77800   1950416      Julian Gaviria

[54589 rows x 2 columns]

#### 下一步就可以把`actor_id_with_names`與前面得到的`imdb_score_groupby_genres_and_person_id_df`進行連接，增加name變量，從而展示平均評分最高的演員名字。

In [41]:
imdb_score_groupby_genres_and_person_id_df_names = pd.merge(imdb_score_groupby_genres_and_person_id_df, actor_id_with_names, on="person_id")
imdb_score_groupby_genres_and_person_id_df_names

genres person_id  imdb_score                name
0        action      1000    6.866667        Martin Sheen
1        action    100007    7.000000         Dulce María
2        action    100013    6.400000          Wang Ziwen
3        action    100019    6.500000      Michel Beaudry
4        action    100020    6.500000    Patrice Bélanger
...         ...       ...         ...                 ...
168876  western    993735    6.500000      Peyton Jackson
168877  western    998673    7.300000   Stephen R. Estler
168878  western    998674    7.300000       Grace LeSueur
168879  western    998675    7.300000         Ethan Dubin
168880  western     99940    4.000000  Christopher Kahler

[168881 rows x 4 columns]

#### 列"imdb_score"與列"name"交換位置

In [42]:
imdb_score_groupby_genres_and_person_id_df_names = imdb_score_groupby_genres_and_person_id_df_names[['genres', 'person_id', 'name', 'imdb_score']]
imdb_score_groupby_genres_and_person_id_df_names

genres person_id                name  imdb_score
0        action      1000        Martin Sheen    6.866667
1        action    100007         Dulce María    7.000000
2        action    100013          Wang Ziwen    6.400000
3        action    100019      Michel Beaudry    6.500000
4        action    100020    Patrice Bélanger    6.500000
...         ...       ...                 ...         ...
168876  western    993735      Peyton Jackson    6.500000
168877  western    998673   Stephen R. Estler    7.300000
168878  western    998674       Grace LeSueur    7.300000
168879  western    998675         Ethan Dubin    7.300000
168880  western     99940  Christopher Kahler    4.000000

[168881 rows x 4 columns]

#### 現在針對流派和演員分組的IMDB評分數據已經整理好，可以進入後續的分析步驟了。

#


### **5.深層整理**

#### 我們當前可以繼續做一些數據整理，比如對上面的結果再次進行分組，找出各個流派里演員作品最高的平均評分是多少、最高評分對應的演員名字是什麼。
#### 要得到這一結果，我們需要再次用`genres`進行分組，然後提取出`imdb_score`變量，計算其最大值。

In [43]:
genres_max_scores = imdb_score_groupby_genres_and_person_id_df_names.groupby("genres")["imdb_score"].max()
genres_max_scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

#### 在我們知道最高分後，可以把以上結果和之前得到的`imdb_score_groupby_genres_and_person_id_df_name`再次進行連接，得到最高分對應的各個演員ID是什麼，也就是這個最高平均分是哪位演員拿到的。

In [44]:
genres_max_scores_person_id = pd.merge(imdb_score_groupby_genres_and_person_id_df, genres_max_scores, on=["genres", "imdb_score"])
genres_max_scores_person_id

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

#### 下一步就可以把`actor_id_with_names`與前面得到的`genres_max_score_with_person_id`進行連接，增加`name`變量，從而展示平均評分最高的演員名字。

In [45]:
genres_max_scores_person_id_names = pd.merge(genres_max_scores_person_id, actor_id_with_names, on="person_id")
genres_max_scores_person_id_names

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     22311         8.9     Koichi Yamadera
133  western     28166         8.9  Megumi Hayashibara
134  western     28180         8.9      Unsho Ishizuka
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

#### 為了把相同流派都排序在一起，我們還可以用`sort_values`方法，把結果裡面的行根據`genres`進行排序，然後用`reset_index`把索引重新排序。

In [48]:
genres_max_score_with_person_id_names = genres_max_scores_person_id_names.sort_values("genres")
genres_max_score_with_person_id_names

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
134  western     28180         8.9      Unsho Ishizuka
132  western     22311         8.9     Koichi Yamadera
133  western     28166         8.9  Megumi Hayashibara
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

In [ ]:
###索引重新排序后，DataFrame会多出`index`一列，我们可以再把`index`列进行删除。
###genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values("genres").reset_index().drop("index", axis=1)
###genres_max_score_with_actor_name